knative-operator/pkg/webhook/knativeserving/webhook_validating_test.go

```diff
@@ -80,6 +80,102 @@ func TestLoneliness(t *testing.T) {
+	validator.InjectDecoder(&mockDecoder{ks1})
+	validator.InjectClient(fake.NewFakeClient(mockClusterVersion("4.1.12")))
+	result := validator.Handle(context.TODO(), types.Request{})
+	if result.Response.Allowed {
+		t.Error("The version is too low, but the request is allowed")
+	}
+}
+
+func TestInvalidPrereleaseVersion(t *testing.T) {
+	os.Clearenv()
+	os.Setenv("MIN_OPENSHIFT_VERSION", "4.3.13")
+	validator := KnativeServingValidator{}
+	validator.InjectDecoder(&mockDecoder{ks1})
+	validator.InjectClient(fake.NewFakeClient(mockClusterVersion("4.1.13-alpha")))
+	result := validator.Handle(context.TODO(), types.Request{})
+	if result.Response.Allowed {
+		t.Error("The version is too low, but the request is allowed")
+	}
+}
+
+func TestValidMajorVersion(t *testing.T) {
+	os.Clearenv()
+	os.Setenv("MIN_OPENSHIFT_VERSION", "4.3.0-0")
+	validator := KnativeServingValidator{}
+	validator.InjectDecoder(&mockDecoder{ks1})
+	validator.InjectClient(fake.NewFakeClient(mockClusterVersion("4.3.0")))
+	result := validator.Handle(context.TODO(), types.Request{})
+	if !result.Response.Allowed {
+		t.Error("The version is later, but the request is not allowed")
+	}
+}
+
+func TestValidMinorVersion(t *testing.T) {
+	os.Clearenv()
+	os.Setenv("MIN_OPENSHIFT_VERSION", "4.3.0-0")
+	validator := KnativeServingValidator{}
+	validator.InjectDecoder(&mockDecoder{ks1})
+	validator.InjectClient(fake.NewFakeClient(mockClusterVersion("4.4.0")))
+	result := validator.Handle(context.TODO(), types.Request{})
+	if !result.Response.Allowed {
+		t.Error("The version is later, but the request is not allowed")
+	}
+}
+
+func TestValidPatchVersion(t *testing.T) {
+	os.Clearenv()
+	os.Setenv("MIN_OPENSHIFT_VERSION", "4.3.0-0")
+	validator := KnativeServingValidator{}
+	validator.InjectDecoder(&mockDecoder{ks1})
+	validator.InjectClient(fake.NewFakeClient(mockClusterVersion("4.3.1")))
+	result := validator.Handle(context.TODO(), types.Request{})
+	if !result.Response.Allowed {
+		t.Error("The version is later, but the request is not allowed")
+	}
+}
+
+func TestValidPrereleaseAlphaVersion(t *testing.T) {
+	os.Clearenv()
+	os.Setenv("MIN_OPENSHIFT_VERSION", "4.3.0-0")
+	validator := KnativeServingValidator{}
+	validator.InjectDecoder(&mockDecoder{ks1})
+	validator.InjectClient(fake.NewFakeClient(mockClusterVersion("4.3.0-alpha")))
+	result := validator.Handle(context.TODO(), types.Request{})
+	if !result.Response.Allowed {
+		t.Error("The version is later, but the request is not allowed")
+	}
+}
+
+func TestValidPrereleaseNightlyVersion(t *testing.T) {
+	os.Clearenv()
+	os.Setenv("MIN_OPENSHIFT_VERSION", "4.3.0-0")
+	validator := KnativeServingValidator{}
+	validator.InjectDecoder(&mockDecoder{ks1})
+	validator.InjectClient(fake.NewFakeClient(mockClusterVersion("4.3.0-0.nightly-2020-01-28-050934")))
+	result := validator.Handle(context.TODO(), types.Request{})
+	if !result.Response.Allowed {
+		t.Error("The version is later, but the request is not allowed")
+	}
+}
```

Comment:
nit: The valid tests seem redundant to `TestPreReleaseVersionConstraint` above. Maybe we should fold them into there and make a similar one for invalid tests for less duplicated code?

Reply:
+1. I didn't see that structure after I rebased. Good point!